LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs

```diff
@@ -155,12 +154,165 @@ public async Task<bool> SendEventAsync(LoRaDeviceTelemetry telemetry, Dictionary
+        int receiveAsyncTaskWaitCount = 0;
+
         public async Task<Message> ReceiveAsync(TimeSpan timeout)
+        {
+            var isUsingPendingRequest = true;
+
+            TaskCompletionSource<Message> localPendingReceiveAsync = null;
+
+            await this.receiveAsyncLock.WaitAsync();
+
+            try
+            {
+                if ((localPendingReceiveAsync = this.receiveAsyncTaskCompletionSource) == null)
+                {
+                    localPendingReceiveAsync = this.SetupNewReceiveAsyncTaskCompletionSource(timeout);
+                    isUsingPendingRequest = false;
+                }
+            }
+            finally
+            {
+                this.receiveAsyncTaskWaitCount++;
+                this.receiveAsyncLock.Release();
+            }
+
+            if (isUsingPendingRequest)
+            {
+                Logger.Log(this.devEUI, $"checking cloud to device message for {timeout}, reusing pending request", LogLevel.Debug);
+            }
+
+            using (var cts = new CancellationTokenSource())
+            {
+                var timer = Task.Delay(timeout, cts.Token);
+                var winner = await Task.WhenAny(localPendingReceiveAsync.Task, timer);
+                if (winner == localPendingReceiveAsync.Task)
+                {
+                    // Cancel the timer tasks so that it does not fire
+                    cts.Cancel();
+
+                    Task<Message> singleFinished;
+                    await this.receiveAsyncLock.WaitAsync();
+                    try
+                    {
+                        if (localPendingReceiveAsync == this.receiveAsyncTaskCompletionSource)
+                        {
+                            Logger.Log(this.devEUI, $"task ReceiveAsync returned before timeout", LogLevel.Debug);
+                            singleFinished = this.receiveAsyncTaskCompletionSource.Task;
+
+                            this.receiveAsyncTaskCompletionSource = null;
+                            this.receiveAsyncTaskWaitCount = 0;
+                        }
+                        else
+                        {
+                            singleFinished = null;
+                        }
+                    }
+                    finally
+                    {
+                        this.receiveAsyncLock.Release();
+                    }
+
+                    // Finished can be null if two race for the value of pendingReceiveAsync
+                    // In that case the winner will handle the message
+                    if (singleFinished != null && singleFinished.IsFaulted)
+                    {
+                        Logger.Log(this.devEUI, $"error in task checking cloud to device message: {singleFinished.Exception?.Message}", LogLevel.Error);
+                        return null;
+                    }
+
+                    return singleFinished?.Result;
+                }
+                else
+                {
```

Comment:
nit: can we add a comment here that this is where we lost the race/timed out, and stop listening for callbacks